passes/opt/muxpack.cc

```diff
@@ -0,0 +1,270 @@
+				for (auto bit : sigmap(wire))
+					sigbit_with_non_chain_users.insert(bit);
+			}
+		}
+
+		for (auto cell : module->cells())
+		{
+			if (cell->type.in("$mux", "$pmux") && !cell->get_bool_attribute("\\keep"))
+			{
+				SigSpec a_sig = sigmap(cell->getPort("\\A"));
+				SigSpec b_sig;
+				if (cell->type == "$mux")
+					b_sig = sigmap(cell->getPort("\\B"));
+				SigSpec y_sig = sigmap(cell->getPort("\\Y"));
+   
+				if (sig_chain_next.count(a_sig))
+					for (auto a_bit : a_sig.bits())
+						sigbit_with_non_chain_users.insert(a_bit);
+				else {
+					sig_chain_next[a_sig] = cell;
+					candidate_cells.insert(cell);
+				}
+
+				if (!b_sig.empty()) {
+					if (sig_chain_next.count(b_sig))
+						for (auto b_bit : b_sig.bits())
+							sigbit_with_non_chain_users.insert(b_bit);
+					else {
+						sig_chain_next[b_sig] = cell;
+						candidate_cells.insert(cell);
+					}
+				}
+
+				sig_chain_prev[y_sig] = cell;
+				continue;
+			}
+
+			for (auto conn : cell->connections())
+				if (cell->input(conn.first))
+					for (auto bit : sigmap(conn.second))
+						sigbit_with_non_chain_users.insert(bit);
+		}
+	}
+
+	void find_chain_start_cells()
+	{
+		for (auto cell : candidate_cells)
+		{
+			log_debug("Considering %s (%s)\n", log_id(cell), log_id(cell->type));
+
+			SigSpec a_sig = cell->getPort("\\A");
+			if (cell->type == "$mux") {
+				SigSpec b_sig = cell->getPort("\\B");
+				if (sig_chain_prev.count(a_sig) + sig_chain_prev.count(b_sig) != 1)
+					goto start_cell;
+
+				if (!sig_chain_prev.count(a_sig))
+					a_sig = b_sig;
+			}
+			else if (cell->type == "$pmux") {
+				if (!sig_chain_prev.count(a_sig))
+					goto start_cell;
+			}
+			else log_abort();
+
+			{
+				for (auto bit : a_sig.bits())
+					if (sigbit_with_non_chain_users.count(bit))
+						goto start_cell;
+
+				Cell *c1 = sig_chain_prev.at(a_sig);
+				Cell *c2 = cell;
+
+				if (c1->getParam("\\WIDTH") != c2->getParam("\\WIDTH"))
+					goto start_cell;
+			}
+
+			continue;
+
+		start_cell:
+			chain_start_cells.insert(cell);
+		}
+	}
+
+	vector<Cell*> create_chain(Cell *start_cell)
+	{
+		vector<Cell*> chain;
+
+		Cell *c = start_cell;
+		while (c != nullptr)
+		{
+			chain.push_back(c);
+
+			SigSpec y_sig = sigmap(c->getPort("\\Y"));
+
+			if (sig_chain_next.count(y_sig) == 0)
+				break;
+
+			c = sig_chain_next.at(y_sig);
+			if (chain_start_cells.count(c) != 0)
+				break;
+		}
+
+		return chain;
+	}
+
+	void process_chain(vector<Cell*> &chain)
+	{
+		if (GetSize(chain) < 2)
+			return;
+
+		int cursor = 0;
+		while (cursor < GetSize(chain))
+		{
+			int cases = GetSize(chain) - cursor;
+
+			Cell *first_cell = chain[cursor];
+			dict<int, SigBit> taps_dict;
+
+			if (cases < 2) {
+				cursor++;
+				continue;
+			}
+
+			Cell *last_cell = chain[cursor+cases-1];
+
+			log("Converting %s.%s ... %s.%s to a pmux with %d cases.\n",
+				log_id(module), log_id(first_cell), log_id(module), log_id(last_cell), cases);
```

Comment:
There's no check anywhere here yet that makes sure the control inputs are exclusive. As it is, this pass incorrectly converts the following test case into a `$pmux`:

```
module test ( 
        input wire x, y, z,
        input wire a, b, c, d,
        output reg o
);
        always @* begin
                o = a;
                if (x) o = b;
                if (y) o = c;
                if (z) o = d;
        end
endmodule
```

Reply:
Yes, you're right. I was thinking of a `$pmux` more as a priority mux and not as a one-hot only.
It is concerning that my testing didn't pick this up though:
https://github.com/YosysHQ/yosys/blob/887df8914c64220b9f306b7d21f199fa247224fd/tests/various/muxpack.ys#L1-L15

Reply:
Okay, I've had a stab at creating an `ExclusiveDatabase` that looks through `$eq` cells that compare against a constant, and `$logic_not` cells (which are `$eq` with zero).
For chains of `$pmux` cells it currently suffers from O(N^2) since it has to compare every bit of `S` in one `$pmux` with every bit of `S` on the other...

Currently doesn't handle `$ne` cells for `if (s != 0) if (s != 1) else else` type structures (thus `mux_if_unbal_5_3_invert` test is commented out right now).

Reply:
More explicitly: it will only handle `$eq` cells -- not `$ne` nor `$ge` or any of those...